kratos/utilities/voxel_utilities.cpp

```diff
@@ -0,0 +1,307 @@
+
+            if (Case == 1) {
+                left = 0.5;
+                right = std::min(0.5,min_distance_to_node[i].first);
+            } else if (Case == 2) {
+                left = std::min(0.5,min_distance_to_node[(i+3)%4].second);
+                right = 0.5;
+            } else if (Case == 3) {
+                left = std::min(0.5,min_distance_to_node[(i+3)%4].second);
+                right = std::min(0.5,min_distance_to_node[i].first);
+            }
+            NodePtrType int_left(new Node<3>(1, nodes[i].X() + left*v_left[0], nodes[i].Y() + left*v_left[1], nodes[i].Z() + left*v_left[2]));
+            NodePtrType int_right(new Node<3>(2, nodes[i].X() + right*v_right[0], nodes[i].Y() + right*v_right[1], nodes[i].Z() + right*v_right[2]));
+            NodePtrType c(new Node<3>(2, nodes[i].X() + left*v_left[0] + right*v_right[0], 
+                                                nodes[i].Y() + left*v_left[1] + right*v_right[1], 
+                                                nodes[i].Z() + left*v_left[2] + right*v_right[2]));
+            points(3) = int_left;
+            points(0) = &nodes[i];
+            points(1) = int_right;
+            points(2) = c;
+
+            if (Case != 0) {
+                partial_area =  max_volume -factor*PointsArea(points)/face_area;
+            } else {
+                partial_area = 0;
+            }
+        }
+        area += partial_area;
+    }
+    return area;    
+}
+
+/**********************************************************************************/
+/**********************************************************************************/
+
+double VoxelUtilities::PointsArea(const PointsArrayType& rPoints) {
+    const NodeType& p0 = rPoints[0];
+    const NodeType& p1 = rPoints[1];
+    const NodeType& p2 = rPoints[2];
+    const NodeType& p3 = rPoints[3];
+
+    const array_1d<double,3> u1 = (p1.Coordinates() - p0.Coordinates());
+    const array_1d<double,3> v1 = (p3.Coordinates() - p0.Coordinates());
+
+    const array_1d<double,3> u2 = (p3.Coordinates() - p2.Coordinates());
+    const array_1d<double,3> v2 = (p1.Coordinates() - p2.Coordinates());
+
+    array_1d<double,3> area1;
+    array_1d<double,3> area2;
+
+    MathUtils<double>::CrossProduct(area1,u1,v1);
+    MathUtils<double>::CrossProduct(area2,u2,v2);
+
+    return 0.5*(norm_2(area1) + norm_2(area2));
+}
+
+/**********************************************************************************/
+/**********************************************************************************/
+
+double VoxelUtilities::EdgeFilledPortion(std::vector<double>& Distances, const PointsArrayType& rEnds) {
+    const double length = Distances[Distances.size() - 1];
+    double portion = 0; 
+    bool inside = true;       
+
+    const bool left_inside = rEnds[0].GetSolutionStepValue(DISTANCE) > 0;
+    const bool right_inside = rEnds[1].GetSolutionStepValue(DISTANCE) > 0; 
+    
```

Comment:
I would create a `const` variable for `Distances.size()` as you are using it several times.